examples/Example12/Example12.cpp

```diff
@@ -0,0 +1,268 @@
+  G4VPhysicalVolume *world = parser.GetWorldVolume();
+
+  if (world == nullptr) {
+    std::cerr << "Example12: World volume not set properly check your setup selection criteria or GDML input!\n";
+    return world;
+  }
+
+  // - PHYSICS
+  // --- Create particles that have secondary production threshold.
+  G4Gamma::Gamma();
+  G4Electron::Electron();
+  G4Positron::Positron();
+  G4Proton::Proton();
+  G4ParticleTable *partTable = G4ParticleTable::GetParticleTable();
+  partTable->SetReadiness();
+
+  // - REGIONS
+  if (world->GetLogicalVolume()->GetRegion() == nullptr) {
+    // Add default region if none available
+    auto defaultRegion = new G4Region("DefaultRegionForTheWorld"); // deleted by store
+    defaultRegion->AddRootLogicalVolume(world->GetLogicalVolume());
+    defaultRegion->SetProductionCuts(G4ProductionCutsTable::GetProductionCutsTable()->GetDefaultProductionCuts());
+  }
+  for (auto region : *G4RegionStore::GetInstance())
+    region->UsedInMassGeometry(true); // make sure all regions are marked as used
+
+  // - UPDATE COUPLES
+  std::cout << "updating material-cut couples based on " << G4RegionStore::GetInstance()->size() << " regions ...\n";
+  G4ProductionCutsTable *theCoupleTable = G4ProductionCutsTable::GetProductionCutsTable();
+  theCoupleTable->UpdateCoupleTable(world);
+  return world;
+}
+
+const vecgeom::cxx::VPlacedVolume *InitVecGeom(const std::string &gdml_file, int cache_depth)
+{
+  // Import the gdml file into VecGeom
+  vecgeom::GeoManager::Instance().SetTransformationCacheDepth(cache_depth);
+  vgdml::Parser vgdmlParser;
+  auto middleWare = vgdmlParser.Load(gdml_file.c_str(), false, copcore::units::mm);
+  if (middleWare == nullptr) {
+    std::cerr << "Failed to read geometry from GDML file '" << gdml_file << "'" << std::endl;
+    return nullptr;
+  }
+
+  const vecgeom::VPlacedVolume *world = vecgeom::GeoManager::Instance().GetWorld();
+  if (world == nullptr) {
+    std::cerr << "GeoManager world volume is nullptr" << std::endl;
+    return nullptr;
+  }
+  return world;
+}
+
+int *CreateMCCindex(const G4VPhysicalVolume *g4world, const vecgeom::VPlacedVolume *world,
+                    const G4HepEmState &hepEmState)
+{
+  const int *g4tohepmcindex = hepEmState.fData->fTheMatCutData->fG4MCIndexToHepEmMCIndex;
+
+  // - FIND vecgeom::LogicalVolume corresponding to each and every G4LogicalVolume
+  int nphysical = 0;
+
+  int nvolumes   = vecgeom::GeoManager::Instance().GetRegisteredVolumesCount();
+  int *mcc_index = new int[nvolumes];
+  memset(mcc_index, 0, nvolumes * sizeof(int));
+
+  // recursive geometry visitor lambda matching one by one Geant4 and VecGeom logical volumes
+  // (we need to make sure we set the right MCC index to the right volume)
+  typedef std::function<void(G4LogicalVolume const *, vecgeom::LogicalVolume const *)> func_t;
+  func_t visitAndSetMCindex = [&](G4LogicalVolume const *g4vol, vecgeom::LogicalVolume const *vol) {
```

Comment:
Just `auto`, that's what it was invented for

Reply:
not yet for recursive lambdas unfortunatelly:
```/home/agheata/geant_src/AdePT/examples/Example12/example12.cpp:129:7: error: use of ‘visitAndSetMCindex’ before deduction of ‘auto’
       visitAndSetMCindex(g4pvol->GetLogicalVolume(), pvol->GetLogicalVolume());
       ^~~~~~~~~~~~~~~~~~
```